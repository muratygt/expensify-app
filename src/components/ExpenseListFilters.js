import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
export class ExpenseListFilters extends Component {
	state = {
		calendarFocused: null
	};
	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));
	};

	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value);
	};

	onSortChange = (e) => {
		if (e.target.value === 'date') {
			return this.props.sortByDate();
		} else if (e.target.value === 'amount') {
			return this.props.sortByAmount();
		}
	};

	render() {
		return (
			<div>
				<input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
				<select value={this.props.filters.sortBy} onChange={this.onSortChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
					endDate={this.props.filters.endDate}
					endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					showClearDates={true}
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({
	filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
