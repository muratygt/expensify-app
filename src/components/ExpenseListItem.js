import React, { Component } from 'react';

import { Link } from 'react-router-dom';

const ExpenseItemList = ({ id, description, amount, createdAt }) => (
	<div>
		<Link to={`/edit/${id}`}>
			<h3>{description}</h3>
		</Link>
		<p>
			{amount} - {createdAt}
		</p>
	</div>
);
//If you do not want anything from state and just want to connect and get access to dispatch we do not need to write mapStateToProps inside of the connect func.
export default ExpenseItemList;
