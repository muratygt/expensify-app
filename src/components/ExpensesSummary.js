import React, { Component } from 'react';
import { connect } from 'react-redux';
import selectedExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
	const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
	const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

	return (
		<div>
			{expenseCount > 0 && (
				<p>
					Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
				</p>
			)}
		</div>
	);
};

const mapStateToProps = (state) => {
	const visibleExpenses = selectedExpenses(state.expenses, state.filters);

	return {
		expenseCount: visibleExpenses.length,
		expensesTotal: selectExpensesTotal(visibleExpenses)
	};
};

export default connect(mapStateToProps)(ExpensesSummary);
