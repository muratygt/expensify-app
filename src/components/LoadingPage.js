import React from 'react';

const LoadingPage = () => (
	<div className="loaders">
		<img className="loaders__image" src="/images/loader.gif" />
	</div>
);

export default LoadingPage;
