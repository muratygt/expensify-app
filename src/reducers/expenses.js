//EXPENSES REDUCER//

//We simply create a const to reference the default value of the state
const expensesReducerDefaultState = [];

//We created a function for the expenses reducer. We pass in 2 arguments which are state and action.
export default (state = expensesReducerDefaultState, action) => {
	switch (action.type) {
		case 'ADD_EXPENSE':
			return [ ...state, action.expense ]; //combining 2 arrays with spread operator
		case 'REMOVE_EXPENSE':
			return state.filter(({ id }) => id !== action.id);
		case 'EDIT_EXPENSE':
			return state.map((expense) => {
				if (expense.id === action.id) {
					return {
						...expense,
						...action.updates
					};
				} else {
					return expense;
				}
			});
		default:
			return state;
	}
};
