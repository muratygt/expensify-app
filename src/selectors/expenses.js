import moment from 'moment';
//Get Visible Expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
	return expenses
		.filter((expense) => {
			//created an instance of the moment that is entered
			const createdAtMoment = moment(expense.createdAt);

			//we want to check the start date filter is same or before the created day of the expense
			const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;

			const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
			const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

			return startDateMatch && endDateMatch && textMatch;
		})
		.sort((a, b) => {
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1;
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1;
			}
		});
};
