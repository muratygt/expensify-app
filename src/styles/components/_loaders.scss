.loaders{
    align-items: center;
    display: flex;
    height: 100vh;
    width: 100vw;
    justify-content: center;

}

.loaders__image{
height: 6rem;
width: 6rem;
}