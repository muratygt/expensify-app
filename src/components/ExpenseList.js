import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';

//This allows only show the filtered expenses on UI
import selectedExpenses from '../selectors/expenses';

//We define the component that is the connected version.
export const ExpenseList = (props) => (
	<div>
		{props.expenses.length === 0 ? (
			<p>No Expenses</p>
		) : (
			props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)
		)}
	</div>
);

//We define the things that we want to get from the store with mapStateToProps
const mapStateToProps = (state) => {
	return {
		//These are the props we want to pass in. mapStateToProps is HOC
		expenses: selectedExpenses(state.expenses, state.filters)
	};
};

export default connect(mapStateToProps)(ExpenseList);
