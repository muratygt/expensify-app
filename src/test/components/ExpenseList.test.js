//We will tes the component without connecting to store. So we will pass in our own data from fixtures

import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render ExpenseList with expenses', () => {
	const wrapper = shallow(<ExpenseList expenses={expenses} />);
	expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with expenses with empty array', () => {
	const wrapper = shallow(<ExpenseList expenses={[]} />);
	expect(wrapper).toMatchSnapshot();
});
