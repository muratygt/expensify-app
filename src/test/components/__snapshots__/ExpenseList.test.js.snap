// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseItemList
    amount={195}
    createdAt={0}
    description="Bum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseItemList
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseItemList
    amount={4500}
    createdAt={345600000}
    description="Water Bill"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should render ExpenseList with expenses with empty array 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;
